### PR TITLE
KAFKA-5341; Add UnderMinIsrPartitionCount and per-partition UnderMinIsr metrics (KIP-164)

### PR DESCRIPTION
**Pull request by lindong28**

### CODE REVIEW

**Comment by lindong28** [comment]
@becketqin @ijuma This is probably low priority. Can you take a look when you get time?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6364/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6349/
Test PASSed (JDK 8 and Scala 2.12).

**Review by ewencp — COMMENTED** [review]
@lindong28 This looks good and I did a quick test locally to make sure they work as expected.

Should we also update the documentation [here](http://kafka.apache.org/documentation.html#monitoring) to include the `UnderMinIsrPartitionCount` metric since it's a good health check to alert on?

**Comment by lindong28** [comment]
@ewencp Thanks for your review! Good point. I have updated the ops.html to specify both `UnderMinIsrPartitionCount` and `OfflineLogDirectoryCount`. `OfflineLogDirectoryCount` is also an important metric to monitor for broker health and it is added in KIP-112.

**Comment by lindong28** [comment]
@ewencp And I have verified using jconsole that both metrics can be found with the specified MBean path.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6398/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6383/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6398/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6413/
Test PASSed (JDK 7 and Scala 2.11).

**Review by ewencp — APPROVED** [review]
LGTM, thanks!

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 1]
```diff
@@ -120,6 +129,15 @@ class Partition(val topic: String,
...
     isLeaderReplicaLocal && inSyncReplicas.size < assignedReplicas.size
 
+  def isUnderMinIsr: Boolean = {
+    leaderReplicaIfLocal match {
+      case Some(leaderReplica) =>
+        inSyncReplicas.size < leaderReplica.log.get.config.minInSyncReplicas
+      case None =>
+        false
+    }
+  }
```
Comment (ijuma) [inline core/src/main/scala/kafka/cluster/Partition.scala:139]: This can be written as a one liner in the same style as `isUnderReplicated`.

Comment (ijuma) [reply core/src/main/scala/kafka/cluster/Partition.scala:139]: Oh, I'm wrong, sorry.

**Pull request closed without merging**